ledger/accountdb.go

```diff
@@ -786,6 +878,57 @@ type resourcesData struct {
 	UpdateRound   uint64        `codec:"z"`
 }
 
+func (rd *resourcesData) IsHolding() bool {
+	return (rd.ResourceFlags & resourceFlagsNotHolding) == resourceFlagsHolding
+}
+
+func (rd *resourcesData) IsOwning() bool {
+	return (rd.ResourceFlags & resourceFlagsOwnership) == resourceFlagsOwnership
+}
+
+func (rd *resourcesData) IsEmptyApp() bool {
+	return rd.SchemaNumUint == 0 &&
+		rd.SchemaNumByteSlice == 0 &&
+		len(rd.KeyValue) == 0 &&
+		len(rd.ApprovalProgram) == 0 &&
+		len(rd.ClearStateProgram) == 0 &&
+		len(rd.GlobalState) == 0 &&
+		rd.LocalStateSchemaNumUint == 0 &&
+		rd.LocalStateSchemaNumByteSlice == 0 &&
+		rd.GlobalStateSchemaNumUint == 0 &&
+		rd.GlobalStateSchemaNumByteSlice == 0 &&
+		rd.ExtraProgramPages == 0
+}
+
+func (rd *resourcesData) IsApp() bool {
+	if (rd.ResourceFlags & resourceFlagsEmptyApp) == resourceFlagsEmptyApp {
+		return true
+	}
+	return !rd.IsEmptyApp()
+}
+
+func (rd *resourcesData) IsEmptyAsset() bool {
+	return rd.Amount == 0 &&
+		!rd.Frozen &&
+		rd.Total == 0 &&
+		rd.Decimals == 0 &&
+		!rd.DefaultFrozen &&
+		rd.UnitName == "" &&
+		rd.AssetName == "" &&
+		rd.URL == "" &&
+		rd.MetadataHash == [32]byte{} ||
+		rd.Manager.IsZero() ||
+		rd.Reserve.IsZero() ||
+		rd.Freeze.IsZero() ||
+		rd.Clawback.IsZero()
+}
+
+func (rd *resourcesData) IsAsset() bool {
+	if (rd.ResourceFlags & resourceFlagsEmptyAsset) == resourceFlagsEmptyAsset {
+		return true
+	}
+	return !rd.IsEmptyAsset()
+}
 func (rd *resourcesData) SetAssetParams(ap basics.AssetParams, haveHoldings bool) {
```

Comment:
btw, I think this is error prone. It is probably better setting `rd.ResourceFlags |= resourceFlagsNotHolding` unconditionally and clear this flag in SetAssetHolding

Reply:
I agree. I'll need to think how to re-implement this so that the *persistence* would be taking advantage of the zero for holding, while the normal operation wouldn't ( i.e. maybe two separate variables ?)